Describe name injection; add two new examples (#6357)
tidyverse/dplyr: vignettes/programming.Rmd

[^promise]: In R, arguments are lazily evaluated which means that until you attempt to use, they don't hold a value, just a __promise__ that describes how to compute the value. You can learn more at <https://adv-r.hadley.nz/functions.html#lazy-evaluation>

### Name injection

> Many data masking functions also use dynamic dots, which gives you another useful feature: generating names programmatically by using `:=` instead of `=`. There are two basics forms, as illustrated below with `tibble()`:

Reviewer: Note for myself: We really need to enable glue syntax with `=` so we no longer have to teach this awkward operator. Probably worth it to do this asap rather than wait for the next big tidyeval update?

Author: Yeah, that'd be nice.

*   If you have the name in an env-variable, you can use glue syntax to interpolate 
    in:

    ```{r}
    name <- "susan"
    tibble("{name}" := 2)
    ```
    

> *   If you the name should be derived from a data-variable in an 

Reviewer: ```suggestion
*   If the name should be derived from a data-variable in an 
```

}
```

> If you want to use the names of variables in the output, you can use embrace the variable name on the left-hand side of `:=`:

Reviewer: ```suggestion
If you want to use the name of a variable in the output, you can embrace the variable name on the left-hand side of `:=` with `{{`:
```

  
When you use `...` in this way, make sure that any other arguments start with `.` to reduce the chances of argument clashes; see <https://design.tidyverse.org/dots-prefix.html> for more details.
  
### Creating multiple columns

You can use variable interpolation with `tibble()` to create columns based on the names of an input:

```{r}
quantile_df <- function(x, probs = c(0.25, 0.5, 0.75)) {
  tibble(
    "{{x}}_val" := quantile(x, probs),
    "{{x}}_quant" := probs
  )
}

x <- 1:5
quantile_df(x)
```

This sort of function is useful inside `summarise()` and `mutate()` which allow you to add multiple columns by returning a data frame:

```{r}
df <- tibble(
  grp = rep(1:3, each = 10),
  x = runif(30),
  y = rnorm(30)
)

df %>%
  group_by(grp) %>%
  summarise(quantile_df(x))
```

> Unfortunately it doesn't quite yet work with `across()` because (for complicated reasons) it generates so-called packed columns that are themselves data frames: 

Reviewer: Are we planning recursive unpacking? Or do we consider across results to be unnamed in principle (by the user) and thus df results should be unpacked?

Author: I don't think we have a concrete proposal on the table yet, but it's probably worth considering for this release.

```{r}
df %>%
  group_by(grp) %>%
  summarise(across(x:y, quantile_df))
```

You'll need an additional unpack to get what you want:

```{r}
df %>%
  group_by(grp) %>%
  summarise(across(x:y, quantile_df)) %>%
  tidyr::unpack(x:y)
```

### Allowing the user to control column names

> If you want to give the user the ability to control the generated names, add an additional argument that uses `rlang::caller_arg()` to get the name of the argument to the function:

Reviewer: oh I had not expected this usage. Initially I thought this'd be the role of `englue()`, e.g. we'd do `prefix = englue("{{x}}")`. Should we keep `caller_arg` in the realm of package developers and use `englue()` here as it is more directly targetted towards data analysis users?

Author: Oh I didn't know about (or more likely, had forgotten) `englue()`, I'll use it instead.

Author: Hmmm `englue()` makes more sense in the context of `ensym`, `enquo` etc, which hardly any one should now use. `capture_arg()` seems pretty straight forward to explain.

Reviewer: I see `englue()` on the level of user-facing syntax of tidyeval. The `en` part fits with `enquo()` etc, but that doesn't really need to be mentioned, we can just say that it is a function that provides the same syntax as on the LHS of `:=` (and soon `=`).

Mine needed `englue()` in one of her course IIRC, so it should be occasionally useful to data analysis users. We should probably discourage very complex name LHS and instead recommend creating intermediate variables with `englue()`.

Author: Ok, sounds good.
